crates/matrix-sdk-sled/src/encode_key.rs

```diff
@@ -0,0 +1,158 @@
+    }
+}
+
+impl<T: EncodeKey + ?Sized> EncodeKey for Box<T> {
+    fn encode(&self) -> Vec<u8> {
+        T::encode(self)
+    }
+}
+
+impl EncodeKey for str {
+    fn encode(&self) -> Vec<u8> {
+        [self.as_bytes(), &[ENCODE_SEPARATOR]].concat()
+    }
+}
+
+impl EncodeKey for String {
+    fn encode(&self) -> Vec<u8> {
+        self.as_str().encode()
+    }
+}
+
+impl EncodeKey for DeviceId {
+    fn encode(&self) -> Vec<u8> {
+        self.as_str().encode()
+    }
+}
+
+impl EncodeKey for EventId {
+    fn encode(&self) -> Vec<u8> {
+        self.as_str().encode()
+    }
+}
+
+impl EncodeKey for RoomId {
+    fn encode(&self) -> Vec<u8> {
+        self.as_str().encode()
+    }
+}
+
+impl EncodeKey for TransactionId {
+    fn encode(&self) -> Vec<u8> {
+        self.as_str().encode()
+    }
+}
+
+impl EncodeKey for MxcUri {
+    fn encode(&self) -> Vec<u8> {
+        let s: &str = self.as_ref();
+        s.encode()
+    }
+}
+
+impl EncodeKey for SecretName {
+    fn encode(&self) -> Vec<u8> {
+        let s: &str = self.as_ref();
+        s.encode()
+    }
+}
+
+impl EncodeKey for UserId {
+    fn encode(&self) -> Vec<u8> {
+        self.as_str().encode()
+    }
+}
+
+impl<A, B> EncodeKey for (A, B)
+where
+    A: AsRef<str>,
+    B: AsRef<str>,
+{
+    fn encode(&self) -> Vec<u8> {
+        [
+            self.0.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+            self.1.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+        ]
+        .concat()
+    }
+}
+
+impl<A, B, C> EncodeKey for (A, B, C)
+where
+    A: AsRef<str>,
+    B: AsRef<str>,
+    C: AsRef<str>,
+{
+    fn encode(&self) -> Vec<u8> {
+        [
+            self.0.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+            self.1.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+            self.2.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+        ]
+        .concat()
+    }
+}
+
+impl<A, B, C, D> EncodeKey for (A, B, C, D)
+where
+    A: AsRef<str>,
+    B: AsRef<str>,
+    C: AsRef<str>,
+    D: AsRef<str>,
+{
+    fn encode(&self) -> Vec<u8> {
+        [
+            self.0.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+            self.1.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+            self.2.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+            self.3.as_ref().as_bytes(),
+            &[ENCODE_SEPARATOR],
+        ]
+        .concat()
+    }
+}
+
+impl EncodeKey for StateEventType {
+    fn encode(&self) -> Vec<u8> {
+        self.to_string().encode()
+    }
+}
+
+impl EncodeKey for GlobalAccountDataEventType {
+    fn encode(&self) -> Vec<u8> {
+        self.to_string().encode()
+    }
+}
+
+/* impl EncodeKey for RoomAccountDataEventType {
+    fn encode(&self) -> Vec<u8> {
+        self.to_string().encode()
+    }
+} */
+
+pub fn encode_key_with_usize<A: AsRef<str>>(s: A, i: usize) -> Vec<u8> {
+    // Not portable across architectures - is this right?
```

Comment:
Wdym by "depending on if this needs to be decoded"?

Reply:
This encodes the tuple (A, usize) into a `Vec<u8>`. If we need to perform the reverse operation, decode a `Vec<u8>` back into `(A, usize)`, we might want to consider using an `u32`. This means we fail at the encoding step on 64 bit architectures instead of failing at the decoding step on 32 bit architectures.

That being said, I have not seen that this needs to be decoded, though I have not thoroughly checked.

Reply:
Left a FIXME comment and opened #561 because this isn't as straight-forward as the refactorings in this PR.